contracts/swappers/CurveSwapper.sol

```diff
@@ -0,0 +1,257 @@
+    tokenOutIndex = type(uint).max;
+
+    for (uint256 i = 0; i < len; i++) {
+      if (address(tokens[i]) == tokenIn) {
+        tokenInIndex = i;
+        break;
+      }
+    }
+
+    for (uint256 i = 0; i < len; i++) {
+      if (address(tokens[i]) == tokenOut) {
+        tokenOutIndex = i;
+        break;
+      }
+    }
+
+    require(tokenInIndex < len, 'Wrong tokenIn');
+    require(tokenOutIndex < len, 'Wrong tokenOut');
+  }
+
+  function _getTokensFromMinter(address minter) private view returns(address[] memory) {
+    address[] memory tempTokens = new address[](COINS_LENGTH_MAX);
+    uint256 count = 0;
+    for (uint256 i = 0; i < COINS_LENGTH_MAX; i++) {
+      address coin = _getCoin(minter, i);
+      if (coin == address(0)) {
+        break;
+      }
+      tempTokens[i] = coin;
+      count++;
+    }
+
+    address[] memory foundTokens = new address[](count);
+    for (uint256 j = 0; j < count; j++) {
+      foundTokens[j] = tempTokens[j];
+    }
+
+    return foundTokens;
+  }
+
+  function _getCoin(address minter, uint256 index) private view returns (address) {
+    try ICurveMinter(minter).coins{gas: 6000}(index) returns (address coin) {
+      return coin;
+    } catch {}
+    return address(0);
+  }
+
+  function _convertToInt(uint256 number) private pure returns (int128) {
+    require(number < COINS_LENGTH_MAX, "Wrong token index");
+    int128[5] memory intArr = [int128(0), int128(1), int128(2), int128(3), int128(4)];
+    return intArr[number];
+  }
+
+  function _getMinter(address pool) private view returns (address minter) {
+    (bool success, bytes memory returnData) = address(pool).staticcall(
+      abi.encodeWithSignature("minter()")
+    );
+
+    if (success) {
+      minter = abi.decode(returnData,(address));
+    } else {
+      (bool withFuncCoins, ) = minter.staticcall(
+        abi.encodeWithSignature("coins(uint256)", 0)
+      );
+      if(withFuncCoins) {
+        minter = pool;
+      } else {
+        revert("This pool is not a normal curve type");
+      }
+    }
+  }
+
+  function _callGetDY(
+    address minter, uint256 tokenInIndex, uint256 tokenOutIndex, uint256 dx
+  ) private view returns (uint256 dy) {
+
+    (bool uintSuccess, bytes memory uintReturnData) = minter.staticcall(
+      abi.encodeWithSignature("get_dy(uint256,uint256,uint256)",
+      tokenInIndex, tokenOutIndex, dx
+      )
+    );
+
+    if (uintSuccess) {
+      dy = abi.decode(uintReturnData,(uint256));
+    } else {
+      ( , bytes memory intReturnData) = minter.staticcall(
+        abi.encodeWithSignature("get_dy(int128,int128,uint256)",
+        _convertToInt(tokenInIndex), _convertToInt(tokenOutIndex), dx
+        )
+      );
+      dy = abi.decode(intReturnData,(uint256));
+    }
+  }
+
+  function _callExchange(
+    address minter, uint256 tokenInIndex, uint256 tokenOutIndex, uint256 dx, uint256 minDy
+  ) private returns (uint256 amountOut) {
+
+    (bool uintSuccess, bytes memory uintReturnData) = minter.call(
```

Comment:
is it necessary have two calls?

Reply:
yes
we have two types of exchange function:
1. "exchange(uint256,uint256,uint256,uint256)"
2. "exchange(int128,int128,uint256,uint256)"